smacc/include/smacc/impl/smacc_state_machine_impl.h

```diff
@@ -249,46 +249,93 @@ namespace smacc
     {
       template <typename TSmaccSignal, typename TMemberFunctionPrototype, typename TSmaccObjectType>
       boost::signals2::connection bindaux(TSmaccSignal &signal, TMemberFunctionPrototype callback,
-                                          TSmaccObjectType *object);
+                                          TSmaccObjectType *object, std::shared_ptr<CallbackCounterSemaphore> callbackCounter);
     };
 
     template <>
     struct Bind<1>
     {
       template <typename TSmaccSignal, typename TMemberFunctionPrototype, typename TSmaccObjectType>
-      boost::signals2::connection bindaux(TSmaccSignal &signal, TMemberFunctionPrototype callback, TSmaccObjectType *object)
+      boost::signals2::connection bindaux(TSmaccSignal &signal, TMemberFunctionPrototype callback, TSmaccObjectType *object, std::shared_ptr<CallbackCounterSemaphore> callbackCounter)
       {
-        return signal.connect([=]() { return (object->*callback)(); });
+        return signal.connect(
+
+        [=]()
+        {
+          if(callbackCounter == nullptr)
+          {
+            return (object->*callback)();
+          }
+          else if (callbackCounter->acquire())
+          {
+            (object->*callback)();
+            callbackCounter->release();
```

Comment:
I'm a bit confused here, why does one branch of the if/else have a return and the other not? The lambda's return type is also void, so I'd expect no return is needed at all